### PR TITLE
oAuth does not recognize secret sent from clients using HTTP Basic auth

### PR DESCRIPTION
**Pull request by leleuj**

### CODE REVIEW

File: cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java [hunk 1]
```diff
@@ -81,6 +88,75 @@ private boolean verifyAccessTokenRequest(final HttpServletRequest request) {
...
+            try {
+                credentials = basicAuthExtractor.extract(new J2EContext(request, null));
+            } catch (final TechnicalException e) {
+                logger.error("Cannot get clientId / secret from header", e);
+            }
+        }
+        if (credentials != null) {
+            return credentials;
+        } else {
+            return new UsernamePasswordCredentials(null, null, null);
```
Comment (mmoayyed) [inline cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java]: Could we return null instead? I think semantically, that makes more sense than initializing the credentials with null values.

Comment (leleuj) [inline cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java]: I agree: that's not "pretty", but it saves `null` check on many places. My first design was to return `null`, the ideal would be `Optional` in fact.
I can revert if you want...

Comment (mmoayyed) [inline cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java]: I like optional best! Let's get java8 into master, and then you can massage it that way. I am working on tests and should have an update within the hour or so if not earlier.

Comment (leleuj) [inline cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java]: OK. Let's do that, I will wait for Java 8.

Comment (mmoayyed) [inline cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20AccessTokenController.java]: Done. You should be able to merge and continue.

**Comment by leleuj** [comment]
@mmoayyed : I just updated the pull request to use the Java 8 `Optional` feature.

**Comment by mmoayyed** [comment]
Rock and roll :) :+1:

**Comment by leleuj** [comment]
There are failed tests which are not related to my changes and that I don't reproduce locally.

**Comment by leleuj** [comment]
```
org.jasig.cas.authentication.FileTrustStoreSslSocketFactoryTests > verifyTrustStoreLoadingSuccessfullyWithCertAvailable2 FAILED
    java.lang.AssertionError
        at org.junit.Assert.fail(Assert.java:86)
```

**Comment by mmoayyed** [comment]
Ignore that for now please. The endpoint used by that test appears to be down.

**Comment by leleuj** [comment]
OK. Good. So this pull request is ready for merge.

**Pull request merged**